bundle/src/main/java/com/adobe/acs/commons/properties/impl/PropertyAggregatorServiceImpl.java

```diff
@@ -0,0 +1,273 @@
+        } catch (IntrospectionException e) {
+            log.error("IntrospectionException", e);
+        } catch (IllegalAccessException e) {
+            log.error("IllegalAccessException", e);
+        } catch (InvocationTargetException e) {
+            log.error("InvocationTargetException", e);
+        }
+    }
+
+    /**
+     * Add the properties of a page to the given map.  Excluded properties are found in the configuration for this service.
+     * @param targetPropertyMap the map that should be updated with the properties and their values
+     * @param page the page containing properties
+     */
+    private void addPagePropertiesToMap(final Map<String,Object> targetPropertyMap, final Page page, final String prefix) {
+        ValueMap pageProperties = page.getProperties();
+        Set<Map.Entry<String, Object>> entries = pageProperties.entrySet();
+        for (Map.Entry<String, Object> entry : entries) {
+            if (notInMap(targetPropertyMap, entry.getKey())
+                    && isNotExcluded(entry.getKey())
+                    && isAllowedType(pageProperties.get(entry.getKey()))) {
+                targetPropertyMap.put(prefix(prefix, entry.getKey()), pageProperties.get(entry.getKey()));
+            }
+        }
+    }
+
+    /**
+     * Recursively traverse the properties of a page and it's parents until that parent is not a page or is null, which
+     * means it's the root, and add them to the given map.  The traversing will stop when the parent is a node that is
+     * not a page, such as a folder.  This is necessary to support things such as editable templates under `/conf`.
+     *
+     * @param targetPropertyMap the map that should be updated with the properties and their values
+     * @param page the page containing properties
+     */
+    private void addInheritedPagePropertiesToMap(final Map<String,Object> targetPropertyMap, final Page page) {
+
+        addPagePropertiesToMap(targetPropertyMap, page, INHERITED_PAGE_PROP_PREFIX);
+
+        Resource pageResource = page.adaptTo(Resource.class);
+        if (pageResource != null) {
+            Resource parentResource = pageResource.getParent();
+
+            if (parentResource != null && NameConstants.NT_PAGE.equalsIgnoreCase(parentResource.getResourceType())) {
+                addInheritedPagePropertiesToMap(targetPropertyMap, parentResource.adaptTo(Page.class));
+            }
+        }
+    }
+
+    /**
+     * Add additional data from custom Java beans that are defined in the configuration for this service to the given map.
+     * @param targetPropertyMap the map that should be updated with the properties and their values
+     * @param page the page for the context of the bean properties
+     */
+    private void addAdditionalDataToMap(final Map<String,Object> targetPropertyMap, final Page page) {
+        if (!additionalData.isEmpty()) {
+            for (Map.Entry<String, String> entry : additionalData.entrySet()) {
```

Comment:
You should add to the documentation, that this feature is supposed to be used only rarely. Adding lots of classes here are likely to create a huge overhead, as all of them are tried for each and every ancestor. 
Also these model classes should be built quite resilient, as they are tried to get adapted no matter if they match.

Reply:
That's actually the point of the generic Exception catching - I don't want custom code to ever break rendering of a page because they implemented their sling model incorrectly